### ISSUE TITLE
Clarifying terminology in gl:: depth (or, 'depth read' versus 'depth test')

### ISSUE THREAD

**Issue by richardeakin** [cinder/Cinder#1071]
I recently found the following (from [here](https://www.opengl.org/wiki/GLAPI/glDepthFunc#Notes), and in more general [here](https://www.opengl.org/wiki/Depth_Test)) info when trying to do some more complex depth + blending:

> Even if the depth buffer exists and the depth mask is non-zero, the depth buffer is not updated if the depth test is disabled. In order to unconditionally write to the depth buffer, the depth test should be enabled and set to GL_ALWAYS (see glDepthFunc​).

So, I take that to mean is that the word 'test' is actually a better fit rather than 'read', since you can't write unless you `glEnable( GL_DEPTH_TEST )` as well. We made the name changes to use 'read' and 'write' instead of 'test' and 'mask' because they are easier to remember, but this was somewhat of an oversite.  The effected API's that I know of are:

* The arguments to `ScopedDepth` ([here](https://github.com/cinder/Cinder/blob/master/include/cinder/gl/scoped.h#L244-L248)).
* `gl::enableDepthRead()` (as opposed to the prospective `gl::enableDepthTest()`).

I'll also note that `ScopedDepth( false, true )`, meaning disable read but unable write, doesn't currently work - nothing gets written to the z buffer as `GL_DEPTH_TEST` is off.

On a related note, is whether we should rename depth 'write' calls to use the word 'mask' instead, to match GL. I think this one is more questionable, as [opengl docs](https://www.opengl.org/wiki/GLAPI/glDepthMask) themselves almost use the words interchangebly (" enable or disable writing into the depth buffer"), but we're soliting the opinion of others since we have the opportunity to make the name change now if that is decision.

**Comment by richardeakin** [comment]
We talked about it here and came up with the following solution for how `gl::ScopedDepth*` should work:

```cpp
gl::ScopedDepth depthScope( true ); // enable test and write, doesn't touch depth func
gl::ScopedDepth depthScope( false ); // disables test and write, doesn't touch depth func
gl::ScopedDepth depthScope( true, GL_LESS ); // enable test and write with scoped depth function as well

gl::ScopedDepthTest depthTestScope( true ); // enables test, doesn't touch write or depth func
gl::ScopedDepthTest depthTestScope( true, GL_LESS ); // enables test and sets depth func, doesn't touch write

gl::ScopedDepthWrite depthWriteScope( true ); // enables write, doesn't touch test or depth func
```

And then the freestanding functions become:
```cpp
gl::enableDepthTest();
gl::enableDepthWrite(); // does not enable test - we should document this as test is needed for write to work
```

Opinions are welcome, else we'll implement these changes in the near future.

**Comment by richardeakin** [comment]
I noticed there is also `gl::enableStencilRead()`, it feels like we should rename that to `gl::enableStencilTest()` as well.

I'm also going to fix #792 while in here and implement `enableStencilWrite()` in terms of `gl::context()->stencilMask()`.

**Comment by richardeakin** [comment]
Hm, oversite on the enableStencilWrite() situation - details on #792. I think we should just remove it and the disable() counterpart.

**Comment by paulhoux** [comment]
Rich and I discussed this a bit just now, but if I am honest I would rather stick to a single ```ScopedDepth``` wrapper that supports the following:

```
gl::ScopedDepth depthScope( true ); // enables reading from and writing to the Z-buffer
gl::ScopedDepth depthScope( false ); // disables reading from and writing to the Z-buffer

gl::ScopedDepth depthScope( true, false ); // enables reading from, disables writing to Z-buffer (e.g. when rendering transparent objects )
gl::ScopedDepth depthScope( false, true ); // equivalent to depthScope( false, GL_ALWAYS )

gl::ScopedDepth depthScope( true, GL_LESS ); // enables reading from, specifies write function.
gl::ScopedDepth depthScope( false, GL_LESS ); // disables reading from, specifies write function.
```

In order to avoid compiler errors where ```GL_LESS``` is converted to a ```bool```, maybe it helps to make the constructors explicit?

-Paul

**Comment by richardeakin** [comment]
No, it doesn't help to use explicit constructors as the GLenum is a preprocessor macro.

**Comment by andrewfb** [comment]
Addressed by #1123
